bundles/org.openhab.binding.bondhome/src/main/java/org/openhab/binding/bondhome/internal/api/BPUPListener.java

```diff
@@ -0,0 +1,282 @@
+            this.socket = null;
+        }
+    }
+
+    private void sendBPUPKeepAlive() {
+        // Create a buffer and packet for the response
+        byte[] buffer = new byte[256];
+        DatagramPacket inPacket = new DatagramPacket(buffer, buffer.length);
+
+        DatagramSocket sock = this.socket;
+        if (sock != null) {
+            logger.trace("Sending keep-alive request ('\\n')");
+            try {
+                byte[] outBuffer = { (byte) '\n' };
+                InetAddress inetAddress = InetAddress.getByName(bridgeHandler.getBridgeIpAddress());
+                DatagramPacket outPacket = new DatagramPacket(outBuffer, 1, inetAddress, BOND_BPUP_PORT);
+                sock.send(outPacket);
+                sock.receive(inPacket);
+                BPUPUpdate response = transformUpdatePacket(inPacket);
+                if (response != null) {
+                    if (!response.bondId.equalsIgnoreCase(bridgeHandler.getBridgeId())) {
+                        logger.warn("Response isn't from expected Bridge!  Expected: {}  Got: {}",
+                                bridgeHandler.getBridgeId(), response.bondId);
+                    } else {
+                        bridgeHandler.setBridgeOnline(inPacket.getAddress().getHostAddress());
+                        numberOfKeepAliveTimeouts = 0;
+                    }
+                }
+            } catch (SocketTimeoutException e) {
+                numberOfKeepAliveTimeouts++;
+                logger.trace("BPUP Socket timeout, number of timeouts: {}", numberOfKeepAliveTimeouts);
+                if (numberOfKeepAliveTimeouts > 10) {
+                    bridgeHandler.setBridgeOffline(ThingStatusDetail.COMMUNICATION_ERROR,
+                            "@text/offline.comm-error.timeout");
+                }
+            } catch (IOException e) {
+                logger.debug("One exception has occurred", e);
+            }
+        }
+    }
+
+    private void receivePackets() {
+        try {
+            DatagramSocket s = new DatagramSocket(null);
+            s.setSoTimeout(SOCKET_TIMEOUT_MILLISECONDS);
+            s.bind(null);
+            socket = s;
+            logger.debug("Listener created UDP socket on port {} with timeout {}", s.getPort(),
+                    SOCKET_TIMEOUT_MILLISECONDS);
+        } catch (SocketException e) {
+            logger.debug("Listener got SocketException", e);
+            datagramSocketHealthRoutine();
+        }
+
+        // Create a buffer and packet for the response
+        byte[] buffer = new byte[256];
+        DatagramPacket inPacket = new DatagramPacket(buffer, buffer.length);
+
+        DatagramSocket sock = this.socket;
+        while (sock != null && !this.shutdown) {
+            // Check if we're due to send something to keep the connection
+            long now = System.currentTimeMillis();
+            long timePassedFromLastKeepAlive = now - timeOfLastKeepAlivePacket;
```

Comment:
Consider `Instant`, as being less susceptible to changes in system time.

Reply:
I'm not a Java pro, but I'm learning. Why would `Instant.now` be any less susceptible to changes in system time? Wouldn't it also shift if the computer's clock changes? My quick googling is that for a monotonic clock in Java when all you want to know is elapsed time, I should be using [System.nanoTime()](https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#nanoTime()).

Reply:
Hmm, it seems you are right and I'm assuming non-existing magic in `Instant`. Please disregard.